source/extensions/clusters/dynamic_forward_proxy/cluster.h

```diff
@@ -61,8 +65,38 @@ class Cluster : public Upstream::BaseDynamicClusterImpl,
     Upstream::HostConstSharedPtr peekAnotherHost(Upstream::LoadBalancerContext*) override {
       return nullptr;
     }
+    absl::optional<Upstream::SelectedPoolAndConnection>
+    selectPool(Upstream::LoadBalancerContext* context, const Upstream::Host& host,
+               std::vector<uint8_t>& hash_key) override;
+    OptRef<Envoy::Http::ConnectionPool::ConnectionLifetimeCallbacks> lifetimeCallbacks() override;
+
+    // Envoy::Http::ConnectionPool::ConnectionLifetimeCallbacks
+    void onConnectionOpen(Envoy::Http::ConnectionPool::Instance& pool,
+                          std::vector<uint8_t>& hash_key,
+                          const Network::Connection& connection) override;
+
+    void onConnectionDraining(Envoy::Http::ConnectionPool::Instance& pool,
+                              std::vector<uint8_t>& hash_key,
+                              const Network::Connection& connection) override;
 
   private:
+    struct ConnectionInfo {
+      Envoy::Http::ConnectionPool::Instance* pool_; // Not a ref to allow assignment in remove().
+      const Network::Connection* connection_;       // Not a ref to allow assignment in remove().
+    };
+    struct LookupKey {
+      const std::vector<uint8_t> hash_key_;
+      const Network::Address::Instance& peer_address_;
+      bool operator==(const LookupKey& rhs) const {
+        return std::tie(hash_key_, peer_address_) == std::tie(rhs.hash_key_, rhs.peer_address_);
+      }
+    };
+    struct LookupKeyHash {
+      size_t operator()(const LookupKey& /*hash*/) const { return 42; }
```

Comment:
try return std::hash(key.peer_address_.asString())?

Reply:
Ah! Good call. (Slightly different syntax seemed to be required to make it work, but that works great)